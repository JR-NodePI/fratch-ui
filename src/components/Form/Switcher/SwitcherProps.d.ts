import { ReactNode } from 'react';

import { IconPlus as Icon } from '../../Icons/Icons';

export type SwitcherProps = {
  value?: boolean;
  className?: string;
  labelLeft?: ReactNode;
  labelRight?: ReactNode;
  title?: string;
  disabled?: boolean;
  onChange?: (switchOn: boolean) => void;
  IconOn?: typeof Icon;
  IconOff?: typeof Icon;
};
