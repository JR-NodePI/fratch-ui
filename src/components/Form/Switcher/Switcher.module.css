.switcher {
  display: flex;
  align-items: center;
  gap: 8px;

  --ft-switcher-width: 60px;
  --ft-switcher-height: calc(var(--ft-switcher-width) / 2 + 2px);
}
.switcher.disabled {
  user-select: none;

  filter: saturate(0.2) brightness(130%) contrast(70%);
}
.switcher .switcher_button {
  border: none;
  width: var(--ft-switcher-width);
  min-width: var(--ft-switcher-width);
  height: var(--ft-switcher-height);
  border-radius: calc(var(--ft-switcher-height) / 2);
  position: relative;
  box-shadow: var(--ft-box-shadow-1);
  border: var(--ft-form-item-border);
  box-sizing: border-box;
}
.switcher .switcher_button .switcher_button_dot {
  top: 2px;
  left: 2px;
  content: '';
  position: absolute;
  width: calc(var(--ft-switcher-width) / 2 - 4px);
  height: calc(var(--ft-switcher-height) - 6px);
  border-radius: calc(var(--ft-switcher-height) / 2 - 2px);
  transition: all 0.3s;
  box-shadow:
    var(--ft-box-shadow-1),
    inset var(--ft-box-shadow-3);
  display: flex;
  align-items: center;
  justify-content: center;
}
.switcher .switcher_button .switcher_button_icon {
  width: 70%;
  height: 70%;
  pointer-events: none;
}
.switcher.on .switcher_button .switcher_button_dot {
  background-color: var(--ft-brand-color);
  left: calc(var(--ft-switcher-width) / 2);
}
.switcher.on .switcher_button .switcher_button_icon {
  color: var(--ft-color-white);
}
.switcher.off .switcher_button .switcher_button_dot {
  background-color: var(--ft-color-grey-2);
}
.switcher.off .switcher_button .switcher_button_icon {
  color: var(--ft-color-text);
}
