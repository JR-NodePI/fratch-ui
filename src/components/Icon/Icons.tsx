export { default as IconArrowDown } from './IconArrowDown';
export { default as IconClose } from './IconClose';
export { default as IconPlus } from './IconPlus';
export { default as IconCheck } from './IconCheck';
