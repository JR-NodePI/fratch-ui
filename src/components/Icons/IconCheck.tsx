import IconContainer from './components/IconContainer';
import { IconProps } from './components/IconProps';

const IconCheck = ({
  type,
  className,
}: Pick<IconProps, 'type' | 'className'>): JSX.Element => {
  return (
    <IconContainer className={className} type={type}>
      {({ iconClassName }) => (
        <svg
          className={iconClassName}
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="-1 -3 18 18"
        >
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2.2"
            d="M1 5.917 5.724 10.5 15 1.5"
          />
        </svg>
      )}
    </IconContainer>
  );
};

export default IconCheck;
