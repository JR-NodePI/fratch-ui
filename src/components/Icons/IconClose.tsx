import IconContainer from './components/IconContainer';
import { IconProps } from './components/IconProps';

const IconClose = ({
  type,
  className,
}: Pick<IconProps, 'type' | 'className'>): JSX.Element => {
  return (
    <IconContainer className={className} type={type}>
      {({ iconClassName }) => (
        <svg
          className={iconClassName}
          aria-hidden="true"
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="-1 -1 16 16"
        >
          <path
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
          />
        </svg>
      )}
    </IconContainer>
  );
};

export default IconClose;
