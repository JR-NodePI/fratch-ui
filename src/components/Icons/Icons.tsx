export { default as IconArrowDown } from './Icons/IconArrowDown';
export { default as IconCheck } from './Icons/IconCheck';
export { default as IconClose } from './Icons/IconClose';
export { default as IconDownload } from './Icons/IconDownload';
export { default as IconEdit } from './Icons/IconEdit';
export { default as IconError } from './Icons/IconError';
export { default as IconExport } from './Icons/IconExport';
export { default as IconImport } from './Icons/IconImport';
export { default as IconInfo } from './Icons/IconInfo';
export { default as IconPause } from './Icons/IconPause';
export { default as IconPin } from './Icons/IconPin';
export { default as IconPlay } from './Icons/IconPlay';
export { default as IconPlus } from './Icons/IconPlus';
export { default as IconSuccess } from './Icons/IconSuccess';
export { default as IconUpload } from './Icons/IconUpload';
export { default as IconVerticalDots } from './Icons/IconVerticalDots';
export { default as IconWarning } from './Icons/IconWarning';
