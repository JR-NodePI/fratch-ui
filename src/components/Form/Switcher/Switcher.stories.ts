import type { Meta, StoryObj } from '@storybook/react';

import { IconCheck, IconClose } from '../../Icons';
import * as Icons from '../../Icons';
import Switcher from './Switcher';

const IconOptions = { Default: undefined, ...Icons };

const meta = {
  title: 'Components/Form/Switcher',
  component: Switcher,
  argTypes: {
    IconOff: {
      options: Object.keys(IconOptions),
      mapping: IconOptions,
      control: { type: 'select' },
    },
    IconOn: {
      options: Object.keys(IconOptions),
      mapping: IconOptions,
      control: { type: 'select' },
    },
    className: { table: { disable: true } },
  },
} satisfies Meta<typeof Switcher>;

export default meta;
type Story = StoryObj<typeof Switcher>;

export const _Switcher: Story = {
  args: {
    title: 'Switch to the left or right',
    labelLeft: 'Label left',
    labelRight: 'Label right',
    IconOff: IconClose,
    IconOn: IconCheck,
  },
};
