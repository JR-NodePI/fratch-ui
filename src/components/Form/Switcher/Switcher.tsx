import { ReactNode, useEffect, useState } from 'react';

import { c } from '../../../helpers/classNameHelpers';
import { IconPlus as Icon } from '../../Icons/Icons';

import styles from './Switcher.module.css';

type SwitcherProps = {
  value?: boolean;
  className?: string;
  labelLeft?: ReactNode;
  labelRight?: ReactNode;
  title?: string;
  disabled?: boolean;
  onChange?: (switchOn: boolean) => void;
  IconOn?: typeof Icon;
  IconOff?: typeof Icon;
};

const Switcher = ({
  value,
  className,
  labelLeft,
  labelRight,
  title,
  disabled,
  onChange,
  IconOn,
  IconOff,
}: SwitcherProps) => {
  const [switchOn, setSwitchOn] = useState<boolean>(value ?? false);

  useEffect(() => {
    setSwitchOn(value ?? false);
  }, [value]);

  const handleToggle = () => {
    if (disabled) return;
    setSwitchOn(!switchOn);
    onChange?.(!switchOn);
  };

  return (
    <label
      title={title}
      className={c(
        styles.switcher,
        switchOn ? styles.on : styles.off,
        disabled ? styles.disabled : '',
        className
      )}
    >
      {labelLeft}
      <button onClick={handleToggle} className={c(styles.switcher_button)}>
        <span className={c(styles.switcher_button_dot)}>
          {switchOn && IconOn && (
            <IconOn className={c(styles.switcher_button_icon)} />
          )}
          {!switchOn && IconOff && (
            <IconOff className={c(styles.switcher_button_icon)} />
          )}
        </span>
      </button>
      {labelRight}
    </label>
  );
};

export default Switcher;
