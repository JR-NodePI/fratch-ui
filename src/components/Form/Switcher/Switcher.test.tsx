import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { IconCheck } from '../../Icons/Icons';
import Switcher from './Switcher';
import { type SwitcherProps } from './SwitcherProps';

describe('Switcher', () => {
  const getComponent = (props: SwitcherProps = {}) => <Switcher {...props} />;
  const setup = (props = {}) => render(getComponent(props));

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render with IconOff', () => {
    const { container } = setup({
      IconOff: IconCheck,
      value: false,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with IconOn', () => {
    const { container } = setup({
      IconOn: IconCheck,
      value: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with label left', () => {
    const { container } = setup({
      labelLeft: 'mock label left',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with label right', () => {
    const { container } = setup({
      labelRight: 'mock label right',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with title', () => {
    const { container } = setup({
      title: 'mock title',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with className', () => {
    const { container } = setup({
      className: 'mock-class-name',
    });
    expect(container).toMatchSnapshot();
  });
});
