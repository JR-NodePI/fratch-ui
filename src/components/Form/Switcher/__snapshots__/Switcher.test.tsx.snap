// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switcher > should render properly 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64"
  >
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      />
    </button>
  </label>
</div>
`;

exports[`Switcher > should render with IconOff 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64"
  >
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      >
        <svg
          class="_icon_2be8fc _primary_2be8fc _switcher_button_icon_b3fc64"
          fill="none"
          viewBox="-1 -3 18 18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 5.917 5.724 10.5 15 1.5"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
          />
        </svg>
      </span>
    </button>
  </label>
</div>
`;

exports[`Switcher > should render with IconOn 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _on_b3fc64"
  >
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      >
        <svg
          class="_icon_2be8fc _primary_2be8fc _switcher_button_icon_b3fc64"
          fill="none"
          viewBox="-1 -3 18 18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 5.917 5.724 10.5 15 1.5"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
          />
        </svg>
      </span>
    </button>
  </label>
</div>
`;

exports[`Switcher > should render with className 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64 mock-class-name"
  >
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      />
    </button>
  </label>
</div>
`;

exports[`Switcher > should render with label left 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64"
  >
    mock label left
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      />
    </button>
  </label>
</div>
`;

exports[`Switcher > should render with label right 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64"
  >
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      />
    </button>
    mock label right
  </label>
</div>
`;

exports[`Switcher > should render with title 1`] = `
<div>
  <label
    class="_switcher_b3fc64 _off_b3fc64"
    title="mock title"
  >
    <button
      class="_switcher_button_b3fc64"
    >
      <span
        class="_switcher_button_dot_b3fc64"
      />
    </button>
  </label>
</div>
`;
